File: javascript/pig-latin/pig-latin.js
export const translator = {
    vowelSounds: ['a', 'e', 'i', 'o', 'u', 'xr', 'yt'],
    translate: function (phrase) {
        return phrase.replace(/[a-zA-Z]+/g, (word) => this.translateWord(word));
    },
    translateWord: function (word) {
        if (this.vowelSounds.some((sound) => word.startsWith(sound))) {
            return word + 'ay';
        } else {
            const matches = word.match(/^(qu|(?:[^aeiou][^aeiouy]*[^aeiouyq]|[^aeiouq])qu|[^aeiou][^aeiouy]*)(.*)$/i);
            return `${matches[2]}${matches[1]}ay`;
        }
    },
};

File: ruby/pig-latin/pig_latin.rb
module PigLatin
  def self.translate(phrase)
    phrase.gsub(/\w+/) { translate_word(_1) }
  end

  def self.translate_word(word)
    case word
    when /^(?:[aeiou]|xr|yt)/
      "#{word}ay"
    when /^([^aeiouq]*qu|[^aeiou][^aeiouy]*)(.+)$/
      "#{$2}#{$1}ay"
    else
      raise RuntimeError, "Cannot translate #{word}"
    end
  end
end
